snapshotter/metrics_integ_test.go

```diff
@@ -0,0 +1,178 @@
+		return fmt.Errorf("Unable to create client to containerd service at %s, is containerd running?: %v", integtest.ContainerdSockPath, err)
+	}
+	kernelArgs := integtest.DefaultRuntimeConfig.KernelArgs + " 8250.nr_uarts=0 quiet loglevel=1"
+	if _, err = fcClient.CreateVM(ctx, &proto.CreateVMRequest{
+		VMID:       vmID,
+		KernelArgs: kernelArgs,
+		RootDrive: &proto.FirecrackerRootDrive{
+			HostPath: "/var/lib/firecracker-containerd/runtime/rootfs-stargz.img",
+		},
+		NetworkInterfaces: []*proto.FirecrackerNetworkInterface{
+			{
+				AllowMMDS: true,
+				CNIConfig: &proto.CNIConfiguration{
+					NetworkName:   "fcnet",
+					InterfaceName: "veth0",
+				},
+			},
+		},
+		MachineCfg: &proto.FirecrackerMachineConfiguration{
+			VcpuCount:  1,
+			MemSizeMib: 512,
+		},
+		ContainerCount: 1,
+	}); err != nil {
+		return fmt.Errorf("Failed to create microVM[%s]: %v", vmID, err)
+	}
+
+	if _, err = fcClient.SetVMMetadata(ctx, &proto.SetVMMetadataRequest{
+		VMID:     vmID,
+		Metadata: fmt.Sprintf(dockerMetadataTemplate, "ghcr.io", noAuth, noAuth),
+	}); err != nil {
+		return fmt.Errorf("Failed to set metadata on microVM[%s]: %v", vmID, err)
+	}
+
+	image, err := client.Pull(ctx, imageRef,
+		containerd.WithPullUnpack,
+		containerd.WithPullSnapshotter(snapshotterName),
+		containerd.WithImageHandlerWrapper(source.AppendDefaultLabelsHandlerWrapper(imageRef, 10*1024*1024)),
+	)
+	if err != nil {
+		return fmt.Errorf("Failed to pull image on microVM[%s]: %v", vmID, err)
+	}
+	if err = client.ImageService().Delete(ctx, image.Name()); err != nil {
+		return fmt.Errorf("Failed to delete image on microVM[%s]: %v", vmID, err)
+	}
+
+	return nil
+}
+
+type metricsTarget struct {
+	Targets []string          `json:"targets"`
+	Labels  map[string]string `json:"labels"`
+}
+
+func verifyMetricsResponse(t *testing.T, numberOfVms int) {
+	sdResponse, err := http.Get(serviceDiscoveryEndpoint)
+	require.NoError(t, err, "Get service discovery failed")
+	defer sdResponse.Body.Close()
+
+	bytes, err := io.ReadAll(sdResponse.Body)
+	require.NoError(t, err, "Failed to read service discovery response body")
+
+	var metricsTargets []metricsTarget
+	err = json.Unmarshal(bytes, &metricsTargets)
+	require.NoError(t, err, "Failed to unmarshall service discovery response")
+
+	require.Len(t, metricsTargets, numberOfVms)
+
+	// Test pulling individual metrics proxies
+	// uniqueTargets acts as a set to ensure each metrics target is a unique host:port
+	uniqueTargets := make(map[string]struct{})
+	for _, mt := range metricsTargets {
+		uniqueTargets[mt.Targets[0]] = struct{}{}
+		metricsResponse, err := http.Get("http://" + mt.Targets[0] + "/metrics")
+		require.NoError(t, err, "Failed to get metrics proxy")
+		defer metricsResponse.Body.Close()
+
+		mBytes, err := io.ReadAll(io.LimitReader(metricsResponse.Body, 300))
+		require.NoError(t, err, "Failed to read metrics response body")
+		require.NotEmpty(t, mBytes)
+	}
+	require.Len(t, uniqueTargets, numberOfVms)
```

Comment:
If all we're verifying is that the snapshotter metrics endpoint returns some bytes, do we actually need to pull an image?

Reply:
The demux snapshotter needs some snapshot API call in order to populate a remote snapshotter in its cache. If we only launch the VM, then yes, the we could pull from its metrics proxy server, but service discovery would not return anything as the demux snapshotter's cache would be empty.